File: cpp/organism/animal/species/Fox.cpp
//
// Created by Filip on 15/04/2025.
//

#include "Fox.h"

Fox::Fox(Point point, World &world)
    : Animal(point, Constants::Animal::Fox::Symbol,
             Constants::Animal::Fox::Strength,
             Constants::Animal::Fox::Initiative, world) {}
void Fox::reproduce() {
    Point newPosition = world.getFreeSpace(position);
    if (newPosition == Point(-1, -1)) {
        return;
    }

    world.queueOrganismAddition(new Fox(newPosition, world));
}

Point Fox::getNewPosition() {
   std::vector<Point> directions = {Point(0, 1), Point(1, 0), Point(0, -1), Point(-1, 0), Point(1, 1), Point(1, -1), Point(-1, 1), Point(-1, -1)};
    Point newPosition;

    int maxAttempts = directions.size();// Maksymalna liczba prób
    int attempts = 0;

    while (!directions.empty() && attempts < maxAttempts) {
        int index = rand() % directions.size();
        newPosition = position + directions[index];
        directions.erase(directions.begin() + index);// Usuń sprawdzony kierunek
        attempts++;

        if (world.isInBounds(newPosition) && !isStrongerOrganismPresent(newPosition)) {
            return newPosition;
        }
    }

    // Jeśli nie znaleziono odpowiedniego pola, zwróć aktualną pozycję
    return position;
}
// Fox.cpp
bool Fox::isStrongerOrganismPresent(Point position) const {
    return world.isStrongerOrganismAt(position, this->getStrength());
}
Organism *Fox::clone() const {
    return new Fox(*this);// Tworzy kopię obiektu za pomocą konstruktora kopiującego
}

File: java/src/organism/animal/species/Fox.java
package organism.animal.species;

import constants.Constants;
import organism.Organism;
import organism.animal.Animal;
import world.World;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Fox extends Animal {

    public Fox(int x, int y, World world) {
        super(x, y, Constants.Animal.Fox.Symbol, Constants.Animal.Fox.Strength, Constants.Animal.Fox.Initiative, world); // Symbol, strength, and initiative for Fox
    }

    @Override
    protected void reproduce() {
        int[] newPosition = world.findNearestFree(x, y);
        if (newPosition == null) {
            return; // No free space for reproduction
        }
        world.addOrganism(new Fox(newPosition[0], newPosition[1], world));
    }

    @Override
    protected int[] getNewPosition() {
        List<int[]> directions = new ArrayList<>();
        directions.add(new int[]{0, 1});    // Up
        directions.add(new int[]{1, 0});    // Right
        directions.add(new int[]{0, -1});   // Down
        directions.add(new int[]{-1, 0});   // Left
        directions.add(new int[]{1, 1});    // Up-Right
        directions.add(new int[]{1, -1});   // Down-Right
        directions.add(new int[]{-1, 1});   // Up-Left
        directions.add(new int[]{-1, -1});  // Down-Left

        Random rand = new Random();
        while (!directions.isEmpty()) {
            int idx = rand.nextInt(directions.size());
            int[] dir = directions.remove(idx);
            int nx = x + dir[0];
            int ny = y + dir[1];
            if (world.isInside(nx, ny) && !isStrongerOrganismPresent(nx, ny)) {
                return new int[]{nx, ny};
            }
        }
        return new int[]{x, y}; // Stay in place if no valid move is found
    }

    private boolean isStrongerOrganismPresent(int nx, int ny) {
        Organism organism = world.getOrganismAt(nx, ny);
        return organism != null && organism.getStrength() > this.getStrength();
    }
}
